src/titiler/xarray/titiler/xarray/io.py

```diff
@@ -0,0 +1,257 @@
+"""titiler.xarray.io"""
+
+import pickle
+from typing import Any, Callable, Dict, List, Optional, Protocol
+from urllib.parse import urlparse
+
+import attr
+import numpy
+import xarray
+from morecantile import TileMatrixSet
+from rio_tiler.constants import WEB_MERCATOR_TMS
+from rio_tiler.io.xarray import XarrayReader
+
+
+class CacheClient(Protocol):
+    """CacheClient Protocol."""
+
+    def get(self, key: str) -> bytes:
+        """Get key."""
+        ...
+
+    def set(self, key: str, body: bytes) -> None:
+        """Set key."""
+        ...
+
+
+def xarray_open_dataset(  # noqa: C901
```

Comment:
I've moved everything within the `xarray_open_dataset` function to ease the customization and also because I've made som dependencies optional